File: cpp/tests/fences_test.cpp
#include "test_harness.hpp"
#include <atomic>
#include <thread>

// Repeats fences_demo.cpp's publish/subscribe pattern many times and
// checks the consumer always observes the fully-published value. Same
// caveat as FencesTest.java: this is a functional smoke test, not a
// race detector -- it verifies the pattern behaves as documented
// across many fresh runs, not that the fences were strictly necessary
// on this particular hardware.
int main() {
    TestHarness t;
    std::cout << "FencesTest\n";

    constexpr int TRIALS = 50;
    bool allCorrect = true;

    for (int trial = 0; trial < TRIALS; ++trial) {
        std::atomic<int> payload{0};
        std::atomic<bool> ready{false};
        int expected = trial + 1;
        int observed = -1;

        std::thread consumer([&] {
            while (!ready.load(std::memory_order_relaxed)) std::this_thread::yield();
            std::atomic_thread_fence(std::memory_order_acquire);
            observed = payload.load(std::memory_order_relaxed);
        });
        std::thread producer([&] {
            payload.store(expected, std::memory_order_relaxed);
            std::atomic_thread_fence(std::memory_order_release);
            ready.store(true, std::memory_order_relaxed);
        });

        producer.join();
        consumer.join();

        if (observed != expected) allCorrect = false;
    }

    t.check("consumer observed the fully-published payload on all " +
                std::to_string(TRIALS) + " trials", allCorrect);

    return t.summary("FencesTest");
}

File: java/tests/FencesTest.java
import java.lang.invoke.MethodHandles;
import java.lang.invoke.VarHandle;

/**
 * Repeats FencesDemo's publish/subscribe pattern many times and checks
 * the consumer always observes the fully-published values. This is a
 * functional smoke test, not a race detector -- correct code always
 * passes it, but passing it doesn't itself prove the release/acquire
 * fence pairing was necessary (only running under a memory-model
 * checker like jcstress could tell you that). What it does verify: the
 * VarHandle fence API behaves as documented and the pattern works
 * end-to-end across many repeated runs, each in fresh state.
 */
public class FencesTest {

    static int payload;
    static int ready;

    static final VarHandle READY;
    static {
        try {
            READY = MethodHandles.lookup()
                    .findStaticVarHandle(FencesTest.class, "ready", int.class);
        } catch (ReflectiveOperationException e) {
            throw new ExceptionInInitializerError(e);
        }
    }

    static int run() throws InterruptedException {
        TestHarness t = new TestHarness();
        System.out.println("FencesTest");

        final int TRIALS = 50;
        boolean allCorrect = true;

        for (int trial = 0; trial < TRIALS; trial++) {
            payload = 0;
            READY.setOpaque(0);
            final int expected = trial + 1;

            Thread producer = new Thread(() -> {
                payload = expected;
                VarHandle.releaseFence();
                READY.setOpaque(1);
            });
            final int[] observed = {-1};
            Thread consumer = new Thread(() -> {
                while ((int) READY.getOpaque() == 0) Thread.onSpinWait();
                VarHandle.acquireFence();
                observed[0] = payload;
            });

            consumer.start();
            producer.start();
            producer.join();
            consumer.join();

            if (observed[0] != expected) allCorrect = false;
        }

        t.check("consumer observed the fully-published payload on all " + TRIALS + " trials",
                allCorrect);

        return t.summary("FencesTest");
    }

    public static void main(String[] args) throws InterruptedException {
        System.exit(run());
    }
}
